### PR TITLE
OIDC Discovery Provider

### PR DESCRIPTION
**Pull request by azdagron**
Provides a minimal implementation of the OIDC well-known discovery
document and the JWKS URI for JWT validation.

See the README.md for details.

### CODE REVIEW

File: support/oidc-discovery-provider/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+| ------------ | -----------------------------------------------------------------| ----------------------------- |
+| `-config`    | Path on disk to the [HCL Configuration](#hcl-configuration) file | `oidc-discovery-provider.conf` |
+
+
+### HCL Configuration
+
+The configuration file is a **required** by the registrar. It contains
+[HCL](https://github.com/hashicorp/hcl) encoded configurables.
+
+| Key                | Type    | Required? | Description                              | Default |
```
Comment (evan2645) [inline support/oidc-discovery-provider/README.md]: nit: would be nice if this list was ordered by required/optional and/or alphabetical

File: support/oidc-discovery-provider/registration_api.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/proto/spire/api/registration"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const (
+	DefaultRegistrationAPIPollInterval = time.Second * 5
```
Comment (evan2645) [inline support/oidc-discovery-provider/registration_api.go]: This is documented (and enforced in config) to default to 10 seconds
```suggestion
	DefaultRegistrationAPIPollInterval = time.Second * 10
```

File: support/oidc-discovery-provider/registration_api.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		s.log.WithError(err).Warn("Failed to fetch bundle")
+		return
+	}
+
+	s.parseBundle(resp.Bundle)
+}
+
+func (s *RegistrationAPISource) parseBundle(bundle *common.Bundle) {
+	if bundle == nil {
+		s.log.Error("No bundle")
```
Comment (evan2645) [inline support/oidc-discovery-provider/registration_api.go]: Something more indicative of the underlying condition would be nice here

```suggestion
		s.log.Error("Received an empty bundle from the Registration API")
```

File: support/oidc-discovery-provider/source.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package main
```
Comment (evan2645) [inline support/oidc-discovery-provider/source.go]: nit: perhaps this file would be better named `jwks_source.go` ?

File: support/oidc-discovery-provider/workload_api.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/spiffe/go-spiffe/workload"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const (
+	DefaultWorkloadAPIPollInterval = time.Second * 5
```
Comment (evan2645) [inline support/oidc-discovery-provider/workload_api.go]: The documented default for this is 10 seconds
```suggestion
	DefaultWorkloadAPIPollInterval = time.Second * 10
```

File: support/oidc-discovery-provider/workload_api.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func (s *WorkloadAPISource) pollOnce(ctx context.Context, client workload_pb.SpiffeWorkloadAPIClient) {
+	// Ensure the stream gets cleaned up
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := client.FetchJWTBundles(ctx, &workload_pb.JWTBundlesRequest{})
+	if err != nil {
+		s.log.WithError(err).Warn("FetchJWTBundles stream failed")
```
Comment (evan2645) [inline support/oidc-discovery-provider/workload_api.go]: nit: An error in layman terms would be nice here too
```suggestion
		s.log.WithError(err).Warn("Failed to fetch JWKS from the Workload API")
```

File: support/oidc-discovery-provider/workload_api.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		return
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		if err == io.EOF {
+			s.log.Warn("FetchJWTBundles stream closed before bundle received")
+			return
+		}
+		s.log.WithError(err).Warn("FetchJWTBundles stream failed")
```
Comment (evan2645) [inline support/oidc-discovery-provider/workload_api.go]:
```suggestion
		s.log.WithError(err).Warn("Failed to fetch JWKS from the Workload API")
```

File: support/oidc-discovery-provider/workload_api.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	stream, err := client.FetchJWTBundles(ctx, &workload_pb.JWTBundlesRequest{})
+	if err != nil {
+		s.log.WithError(err).Warn("FetchJWTBundles stream failed")
+		return
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		if err == io.EOF {
+			s.log.Warn("FetchJWTBundles stream closed before bundle received")
```
Comment (evan2645) [inline support/oidc-discovery-provider/workload_api.go]:
```suggestion
			s.log.Warn("Workload API stream closed before bundle received")
```

File: support/oidc-discovery-provider/workload_api.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+		s.log.WithError(err).Warn("FetchJWTBundles stream failed")
+		return
+	}
+
+	s.parseBundle(resp.Bundles[s.trustDomainID])
+}
+
+func (s *WorkloadAPISource) parseBundle(bundle []byte) {
+	if bundle == nil {
+		s.log.WithField("trust_domain_id", s.trustDomainID).Error("No bundle for trust domain")
```
Comment (evan2645) [inline support/oidc-discovery-provider/workload_api.go]:
```suggestion
		s.log.WithField("trust_domain_id", s.trustDomainID).Error("No bundle for trust domain in Workload API response")
```

File: support/oidc-discovery-provider/workload_api.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	s.mu.RLock()
+	if s.bundle != nil && bytes.Equal(s.bundle, bundle) {
+		s.mu.RUnlock()
+		return
+	}
+	s.mu.RUnlock()
+
+	jwks := new(jose.JSONWebKeySet)
+	if err := json.Unmarshal(bundle, jwks); err != nil {
+		s.log.WithError(err).Error("Failed to unmarshal trust domain bundle")
```
Comment (evan2645) [inline support/oidc-discovery-provider/workload_api.go]:
```suggestion
		s.log.WithError(err).Error("Failed to parse trust domain bundle received from the Workload API")
```

File: support/oidc-discovery-provider/README.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+| ------------------ | --------| ---------| ----------------------------------------- | ------- |
+| `log_format`       | string  | optional | Format of the logs (either `"TEXT"` or `"JSON"`) | `""` |
+| `log_level`        | string  | required | Log level (one of `"panic"`,`"fatal"`,`"error"`,`"warn"`, `"warning"`,`"info"`,`"debug"`,`"trace"`) | `"info"` |
+| `log_path`         | string  | optional | Path on disk to write the log | |
+| `log_requests`     | bool    | optional | If true, all HTTP requests are logged to the debug channel | false |
+| `domain`           | string  | required | The domain the provider is being served from | |
+| `insecure_addr`    | string  | optional\* | If provided, the provider will serve HTTP (insecure) from the specified address instead of using ACME to serve HTTPS. | false |
```
[the remark this answers is not in the document]
Comment (azdagron) [reply support/oidc-discovery-provider/README.md]: I'll leave it undocumented.

File: support/oidc-discovery-provider/README.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+this provider. `insecure_addr` should only be used to bypass this mode if for
+testing or otherwise if you know what you are doing and accept the risks.
+\*\* - Either `registration_api` or `workload_api` must be defined. The provider relies on one of these two APIs to obtain the public key material used to construct the JWKS document.
+
+#### ACME Section
+
+| Key                | Type    | Required?   | Description                              | Default |
+| ------------------ | --------| ----------- | ----------------------------------------- | ------- |
+| `cache_dir`        | string  | required    | The directory used to cache the ACME-obtained credentials. Disabled if explicitly set to the empty string | `./acme-cache` |
+| `directory_url`    | string  | optional    | The ACME directory URL to use. Uses Let's Encrypt if unset. | |
```
[the remark this answers is not in the document]
Comment (azdagron) [reply support/oidc-discovery-provider/README.md]: 👍

File: conf/server/server.conf [hunk 13]
```diff
@@ -7,6 +7,7 @@ server {
     log_level = "DEBUG"
     upstream_bundle = true
     svid_ttl = "1h"
+    jwt_issuer = "FFFFF"
```
[the remark this answers is not in the document]
Comment (azdagron) [reply conf/server/server.conf]: Whoops. This was just for quick testing and shouldn't have been included. I'll remove.

File: support/oidc-discovery-provider/README.md (hunk 1 above)
Comment (azdagron) [reply support/oidc-discovery-provider/README.md]: 👍

File: support/oidc-discovery-provider/README.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+| `email`            | string  | recommended | The email address used to register with the ACME service | |
```
[the remark this answers is not in the document]
Comment (azdagron) [reply support/oidc-discovery-provider/README.md]: I'll make it required.

File: support/oidc-discovery-provider/README.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+| `socket_path`      | string   | required | Path on disk to the Registration API Unix Domain socket. | |
+| `poll_interval`    | duration | optional | How often to poll for changes to the public key material. | `"10s"` |
+
+#### Workload API Section
+
+| Key                | Type     | Required? | Description                              | Default |
+| ------------------ | -------- | ---------| ----------------------------------------- | ------- |
+| `socket_path`      | string   | required | Path on disk to the Registration API Unix Domain socket. | |
+| `poll_interval`    | duration | optional | How often to poll for changes to the public key material. | `"10s"` |
+| `trust_domain`     | string   | required | Trust domain of the workload. This is used to pick the bundle out of the Workload API response. | |
```
[the remark this answers is not in the document]
Comment (azdagron) [reply support/oidc-discovery-provider/README.md]: It isn't in the JWT bundles response. The only way I think we have of locating the trust domain is by fetching x509 SVIDs and parsing trust domain out of the SPIFFE ID in the response...

File: support/oidc-discovery-provider/config.go [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+
+	switch {
+	case c.ACME == nil:
+		if c.InsecureAddr == "" {
+			return nil, errs.New("acme section must be configured")
+		}
+	case c.InsecureAddr != "":
+		return nil, errs.New("insecure_addr and the acme section are mutually exclusive")
+	case !c.ACME.ToSAccepted:
+		return nil, errs.New("acme.tos_accepted must be set to true")
```
[the remark this answers is not in the document]
Comment (azdagron) [reply support/oidc-discovery-provider/config.go]: hmm, maybe. i'm fine with the suggestion.

File: support/oidc-discovery-provider/handler.go [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	jwksBytes, err := json.MarshalIndent(jwks, "", "  ")
+	if err != nil {
+		http.Error(w, "failed to marshal JWKS", http.StatusInternalServerError)
+		return
+	}
+
+	// Allow clients to cache up to 1 hour
+	// TODO: pick a better age? or no caching at all?
```
[the remark this answers is not in the document]
Comment (azdagron) [reply support/oidc-discovery-provider/handler.go]: none at all sounds good for now.

File: support/oidc-discovery-provider/source.go (hunk 4 above)
Comment (azdagron) [reply support/oidc-discovery-provider/source.go]: sgtm
